azavea/echo-locator
# Add Django linting/formatting/testing to CI

## Overview

Adds Python linting and formatting checks as well as Django tests to CI. Also allows testing linting/formatting/Django tests locally through `scripts/test` as well as reformatting locally through new `scripts/format` file.

### Notes

- This implements [GH Super-Linter ](https://github.com/azavea/operations/commit/b562dbb60659fc03b7fc7940410d0d5770ef54d7) for formatting and linting. Although it can format/lint the entire codebase, I set it to check python code in `src/backend/` exclusively. Right now, this checks every file (except `migrations/*` and `manage.py`) with: `gitleaks`, `jscpd`, `black`, `pylint`, `flake8`, and any `json` with `eslint`. I turned off formatting with `isort` since that threw a handful of errors and wasn't something we normally check, but that could easily be turned back on or conversely we could limit to just `black` and `flake8` for now.
- I debated adding `taui` files to the GH Super-Linter check, but it felt worth waiting until we switch to a different bundler and potentially moving frontend/static assets into `src/` directory.
- After adding `scripts/test` to CI, I originally received an error that Django was unable to find the `database` host to run tests. I brought up `database` in `cibuild` so it could be found by the github actions host, and I believe it should stop at the end of the workflow and doesn't affect staging since it's not creating a new tagged container image. However, this is one area where I still have a fuzzy understanding, so it may be worth editing if there are potential side effects.

## Testing Instructions

 * `scripts/update`
 * Create a formatting error somewhere in `src/backend/`
 * `scripts/test`
 * Confirm formatting fails
 * `scripts/format`
 * `scripts/test`
 * Confirm error is reformatted correctly and linting/formatting/testing passes locally
 * Confirm linting/formatting/testing is correctly running in CI through output


Resolves #429 , #452
